sqlglot/optimizer/simplify.py

```diff
@@ -374,47 +373,88 @@ def absorb_and_eliminate(expression, root=True):
     return expression
 
 
-def propagate_constants(expression, root=True):
+# Type alias for a map of columns to their constant value
+ConstantMapping = t.Dict[exp.Column, t.Tuple[int, exp.Expression]]
+
+
+def _get_constant_mapping(ops: t.Iterable[exp.Expression]) -> ConstantMapping:
     """
-    Propagate constants for conjunctions in DNF:
+    Extract a constant mapping from `ops`
+    """
+    constant_mapping: ConstantMapping = {}
+
+    for op in ops:
+        if isinstance(op, exp.EQ):
+            l, r = op.left, op.right
+
+            # TODO: create a helper that can be used to detect nested literal expressions such
+            # as CAST(123456 AS BIGINT), since we usually want to treat those as literals too
+            if isinstance(l, exp.Column) and isinstance(r, exp.Literal):
+                pass
+            elif isinstance(r, exp.Column) and isinstance(l, exp.Literal):
+                l, r = r, l
+            else:
+                continue
+
+            constant_mapping[l] = (id(l), r)
+
+    return constant_mapping
+
+
+def _replace_column(column: exp.Column, constant_mapping: ConstantMapping) -> None:
+    """
+    Potentially replace `column` with a constant from `constant_mapping`
+    """
+    parent = column.parent
+    column_id, constant = constant_mapping.get(column) or (None, None)
+    if (
+        column_id is not None
+        and constant is not None
+        and id(column) != column_id
+        and not (isinstance(parent, exp.Is) and isinstance(parent.expression, exp.Null))
+    ):
+        column.replace(constant.copy())
+
+
+def propagate_constants(expression: exp.Expression, root: bool = True) -> exp.Expression:
+    """
+    Propagate constants for conjunctions:
 
     SELECT * FROM t WHERE a = b AND b = 5 becomes
     SELECT * FROM t WHERE a = 5 AND b = 5
 
     Reference: https://www.sqlite.org/optoverview.html
+
+    This also propagates constants into projections:
+
+    SELECT a FROM x WHERE a = 1 becomes
+    SELECT 1 FROM x WHERE a = 1
```

Comment:
An example where this transformation doesn't fully preserve the semantics:

```
mysql> create temporary table foo with t(c) as (select '1') select c from t where c = 1;
Query OK, 1 row affected (0.01 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> describe foo;
+-------+------------+------+-----+---------+-------+
| Field | Type       | Null | Key | Default | Extra |
+-------+------------+------+-----+---------+-------+
| c     | varchar(1) | NO   |     |         | NULL  |
+-------+------------+------+-----+---------+-------+
1 row in set (0.00 sec)

mysql> select * from foo;
+---+
| c |
+---+
| 1 |
+---+
1 row in set (0.01 sec)
```

And if you transformed it..

```
mysql> create temporary table bar with t(c) as (select '1') select 1 from t where c = 1;
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select * from bar;
+---+
| 1 |
+---+
| 1 |
+---+
1 row in set (0.01 sec)

mysql> describe bar;
+-------+------+------+-----+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+-------+------+------+-----+---------+-------+
| 1     | int  | NO   |     | 0       | NULL  |
+-------+------+------+-----+---------+-------+
1 row in set (0.00 sec)
```

Reply:
Not sure if this is important for your use-case (or in general lol) though, @barakalon

Reply:
Ah, good find!

Two issues:
1. The alias. Early optimizations should fix this, since they'll add the alias. But we can check for that here too and ensure its the same.
2. The type. Oof, this one is tricky. Perhaps we move this to canonicalize and add a CAST? And perhaps we don't replace if there's any type mismatch, too avoid coercion issues?

Reply:
1. I guess it's a tradeoff.. Checking about aliases here means we don't tight couply with earlier rules in the optimizer pipeline, but introduces complexity. (OTOH I don't remember if we can currently safely run this rule without earlier steps, just thinking out loud)
2. Yeah this is definitely an _oof_ moment. Checking for a type mismatch probably makes more sense if we want to avoid this edge case. Not sure I understood the `CAST` suggestion.

Reply:
1. I think simplify is one of the few rules that can safely be run without prior steps.
2. I mean CAST to ensure the exact same type. E.G. `varchar(1)` vs `varchar(2)`

Reply:
Ah yeah, makes sense.